### PR TITLE
RUMM-832 Only authorized data can be uploaded

### PR DESCRIPTION
**Pull request by ncreated**
### What and why?

📦 As part of tracking consent feature, this PR makes sure that only authorized data can be uploaded. This means:
* all data collected with `.granted` consent,
* all data collected with `.pending` consent if it was then changed to `.granted`.

### How?

As in `dd-sdk-android`, `DataMigrator` type is introduced. There are two migrators implemented:
* `DeleteAllDataMigrator` - deletes all data in given folder,
* `MoveDataMigrator` - moves all data from source directory do destination folder.

Migrators are resolved in `ConsentAwareDataWriter` and their `migrate()` work is dispatched to the read/write queue, which keeps all writes (on collection) and reads (on upload) synchronized on the background thread.

### Review checklist

- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference

### CODE REVIEW

File: Sources/Datadog/Core/Persistence/Files/Directory.swift [hunk 1]
```diff
@@ -43,6 +43,24 @@ internal struct Directory {
...
+        let temporaryDirectory = try Directory(withSubdirectoryPath: "com.datadoghq/\(UUID().uuidString)")
+        _ = try FileManager.default.replaceItemAt(url, withItemAt: temporaryDirectory.url)
+        try? FileManager.default.removeItem(at: temporaryDirectory.url)
+    }
+
+    /// Moves all files from this directory to `destinationDirectory`.
+    func moveAllFiles(to destinationDirectory: Directory) throws {
+        try files().forEach { file in
+            let destinationFileURL = destinationDirectory.url.appendingPathComponent(file.name)
+            try? FileManager.default.moveItem(at: file.url, to: destinationFileURL)
```
Comment (buranmert) [inline Sources/Datadog/Core/Persistence/Files/Directory.swift:61]: i guess there is no real solution to this but assuming this `try` will throw sometimes (similar to that extra `,` problem), maybe we can retry the unsuccessful move operations once again, wdyt?

**Review by buranmert — COMMENTED** [review]
LGTM actually, i just left an open question regarding file operations

File: Sources/Datadog/Core/Persistence/Files/Directory.swift (hunk 1 above)
Comment (mariusc83) [reply Sources/Datadog/Core/Persistence/Files/Directory.swift:61]: We're actually doing this in Android. I think we retry 3 times.

File: Tests/DatadogBenchmarkTests/DataStorage/DataMigratorBenchmarkTests.swift [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    func testMigrating500FilesWithDeleteAllDataMigrator() throws {
+        let directory = try Directory(withSubdirectoryPath: UUID().uuidString)
+        defer { directory.delete() }
+
+        measure(metrics: benchmarkMetrics, options: benchmarkOptions) {
+            createFiles(in: directory, count: 500)
+
+            self.startMeasuring()
+            let migrator = DeleteAllDataMigrator(directory: directory)
```
Comment (mariusc83) [inline Tests/DatadogBenchmarkTests/DataStorage/DataMigratorBenchmarkTests.swift:28]: you do not want to create this before measuring ?

File: Sources/Datadog/Core/Persistence/Files/Directory.swift (hunk 1 above)
Comment (ncreated) [reply Sources/Datadog/Core/Persistence/Files/Directory.swift]: Yup, done ✅. We're now retrying the entire `moveAllFiles()` up to 3 times and retrying each `moveItem(at:to:)` up to 3 times as well.

File: Tests/DatadogBenchmarkTests/DataStorage/DataMigratorBenchmarkTests.swift (hunk 2 above)
Comment (ncreated) [reply Tests/DatadogBenchmarkTests/DataStorage/DataMigratorBenchmarkTests.swift:28]: I've been considering this carefully - creation of migrator is part of the actual migration work, so I consider it rather correct to be included in the benchmarked scope.

**Review by mariusc83 — APPROVED** [review]
LGTM

**Pull request merged**